staging/src/k8s.io/cloud-provider/controllers/node/node_controller_test.go

```diff
@@ -924,9 +925,6 @@ func Test_AddCloudNode(t *testing.T) {
 						},
 					},
 				},
-				Spec: v1.NodeSpec{
```

Comment:
I have the same comment about the tests as I did here https://github.com/kubernetes/kubernetes/pull/95342#discussion_r501941558 -- but I don't think it would be simple at the moment to fix the fake cloud provider and back port it into v1.19. It would be good to at least show some manual validation that this patch fixes the bug.

Reply:
@andrewsykim I think the changes to the unit test already validated the fix (e.g. empty providerID for InstancesV2) and as we don't have a separate function like getProviderID here, maybe we could add more unit tests later to the latest version of cloud node controller?

Reply:
I think the test case we're now missing is when `existingNode` does not have `spec.providerID` already set, and is now set based on the providerID returned from `InstanceMetadata`. Testing that would require the fake cloud provider to return a generated provider ID not derived from what's already set on the node:

https://github.com/kubernetes/kubernetes/blob/c1e5e6a5562bdd968977a36720f56362fbf4aa26/staging/src/k8s.io/cloud-provider/fake/fake.go#L325

Reply:
I think the desired behavior from the fake cloud provider is to return `ProviderID` in `InstanceMetadata` from the node spec if it exists, otherwise it should return some generated  providerID like `fake://<ext-id>` where ext-id would be fetched from the ExtID map:

https://github.com/kubernetes/kubernetes/blob/c1e5e6a5562bdd968977a36720f56362fbf4aa26/staging/src/k8s.io/cloud-provider/fake/fake.go#L71

Reply:
gotcha! I think I can update that and merge with this PR onto release 1.19, previously I thought we want a follow-up PR for that

Reply:
> gotcha! I think I can update that and merge with this PR onto release 1.19, previously I thought we want a follow-up PR for that

Sorry if I wasn't clear -- I actually did meant to do this in a follow-up PR, manual validation that this patch worked on a cluster would have been fine :)

Reply:
By manual validation I mean checking if a node (without a pre-set providerID) can be registered and the providerID set

Reply:
gotcha! I've manually validated that :)

Reply:
also created a follow-up PR: https://github.com/kubernetes/kubernetes/pull/95499

Reply:
thank you!